# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ismayilysfli/FlyRank-ml/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I am choosing Refresh / Content Opportunity Scoring as my lane because it can help a content/SEO team decide which pages should be reviewed first when they do not have enough time to review every page manually.


In [1]:
import os
import subprocess
import pandas as pd

REPO_URL = "https://github.com/ismayilysfli/FlyRank-ml.git"
REPO_DIR = "/content/FlyRank-ml"

if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

os.chdir(REPO_DIR)

print("Working directory:", os.getcwd())
print(
    "Dataset found:",
    os.path.exists("data/raw/content_refresh_anonymized.csv")
)

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
print(f"Dataset ready: {len(df):,} pages")

Working directory: /content/FlyRank-ml
Dataset found: True
Dataset ready: 30,000 pages


## 2. The question: decision, action, cost of a wrong call

My work improves the decision of a content/SEO team about which pages should be reviewed first. A human reviewer can inspect a recommended page and decide whether to refresh, expand, protect, monitor, or leave it unchanged. A wrong recommendation can waste the team's limited time on a less important page, while a more important page may be overlooked.

In [2]:
trend_counts = df["trend_direction"].value_counts()

print("Pages by trend:")
print(trend_counts)

Pages by trend:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64


## 3. Quick look at the data (2-3 real numbers)

The dataset contains 30,000 pages, of which 16,262 (54.2%) are labeled as declining. Reviewing all of these pages manually would be difficult, so prioritization is needed. Search volume has almost no linear correlation with actual impressions (0.001), and declining and growing pages have very similar median word counts (2,909 vs. 2,847.5). This suggests that simple rules based on one feature are not enough, and multiple signals should be considered when deciding which pages to review first.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# 1. How many pages are declining?
declining_count = (df["trend_direction"] == "down").sum()
declining_pct = declining_count / len(df) * 100

# 2. Correlation between search volume and impressions
search_impression_corr = df["search_volume"].corr(df["impressions_90d"])

# 3. Median word count for declining vs growing pages
median_word_count = df.groupby("trend_direction")["word_count"].median()

print(f"Total pages: {len(df):,}")
print(f"Declining pages: {declining_count:,} ({declining_pct:.1f}%)")
print(f"Correlation between search volume and impressions: {search_impression_corr:.3f}")
print()
print("Median word count by trend:")
print(median_word_count)

Total pages: 30,000
Declining pages: 16,262 (54.2%)
Correlation between search volume and impressions: 0.001

Median word count by trend:
trend_direction
down      2909.0
flat      2698.5
new       2239.0
stable    2912.5
up        2847.5
Name: word_count, dtype: float64


## 4. Careful words: what I can and can't claim

I can claim that the available data can be used to identify and rank pages that show signals associated with decline or content opportunity. This ranking can support a human reviewer in deciding which pages to inspect first.

I cannot claim that a recommended page definitely needs to be refreshed or that refreshing it will cause its traffic or ranking to improve. The data is observational, so the output should be treated as decision support rather than an automatic action.

In [4]:
print("Available trend labels:")
print(df["trend_direction"].unique())

print("\nThis notebook observes page performance and trend labels.")
print("It does not contain an experiment showing that a refresh causes recovery.")

Available trend labels:
['down' 'stable' 'new' 'up' 'flat']

This notebook observes page performance and trend labels.
It does not contain an experiment showing that a refresh causes recovery.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.